### ISSUE TITLE
tasks waiting in the queued state may get cancelled immediately on start

### ISSUE THREAD

**Issue by mcwumbly** [cloudfoundry/bosh#1778]
On the `cf-infrastructure` team, we have a few tests that start many bosh tasks at once that deploy several `consul` deployments. Some of these tasks wait in the `queued` state for a fair amount of time.  Occasionally, we see that these tasks get cancelled immediately when they start.

We have a working theory that it's related to a race condition when a task goes from queued to processing.  If the task was queued for more than 90s and the status is checked before the checkpoint time is updated, [it gets cancelled](https://github.com/cloudfoundry/bosh/blob/1442da73507cc1c1f376c8dd4836277003469ade/src/bosh-director/lib/bosh/director/api/controllers/tasks_controller.rb#L164-L174).

See also: [thread on Slack](https://cloudfoundry.slack.com/archives/C02HPPYQ2/p1504894095000256)

Here are the logs for one of the cancelled tasks:

```
nm+dm $ bosh task 46 --debug
Using environment 'https://104.154.165.184:25555' as client 'admin'

Task 46

I, [2017-09-07T23:13:46.853493 #2834] [0x2ad3a7ebd110]  INFO -- TaskHelper: Director Version: 262.3.0
I, [2017-09-07T23:13:46.853554 #2834] [0x2ad3a7ebd110]  INFO -- TaskHelper: Enqueuing task: 46
I, [2017-09-07 23:16:57 #7237] []  INFO -- DirectorJobRunner: Looking for task with task id 46
D, [2017-09-07 23:16:57 #7237] [] DEBUG -- DirectorJobRunner: (0.000186s) SELECT NULL
D, [2017-09-07 23:16:57 #7237] [] DEBUG -- DirectorJobRunner: (0.000229s) SELECT * FROM "tasks" WHERE "id" = 46
I, [2017-09-07 23:16:57 #7237] []  INFO -- DirectorJobRunner: Found task #<Bosh::Director::Models::Task @values={:id=>46, :state=>"timeout", :timestamp=>2017-09-07 23:13:46 UTC, :description=>"create deployment", :result=>nil, :output=>"/var/vcap/store/director/tasks/46", :checkpoint_time=>2017-09-07 23:13:46 UTC, :type=>"update_deployment", :username=>"admin", :deployment_name=>"consul-scale-down-5-to-3", :started_at=>nil, :event_output=>"", :result_output=>"", :context_id=>""}>
I, [2017-09-07 23:16:57 #7237] []  INFO -- DirectorJobRunner: Starting task: 46
I, [2017-09-07 23:16:57 #7237] [task:46]  INFO -- DirectorJobRunner: Creating job
D, [2017-09-07 23:16:57 #7237] [task:46] DEBUG -- DirectorJobRunner: (0.000108s) SELECT NULL
D, [2017-09-07 23:16:57 #7237] [task:46] DEBUG -- DirectorJobRunner: (0.000154s) SELECT * FROM "tasks" WHERE "id" = 46
D, [2017-09-07 23:16:57 #7237] [task:46] DEBUG -- DirectorJobRunner: (0.000164s) SELECT NULL
D, [2017-09-07 23:16:57 #7237] [task:46] DEBUG -- DirectorJobRunner: (0.000049s) BEGIN
D, [2017-09-07 23:16:57 #7237] [task:46] DEBUG -- DirectorJobRunner: (0.000306s) UPDATE "tasks" SET "event_output" = ("event_output" || '{"time":1504826217,"error":{"code":10001,"message":"Task 46 cancelled"}}
') WHERE ("id" = 46)
D, [2017-09-07 23:16:57 #7237] [task:46] DEBUG -- DirectorJobRunner: (0.004252s) COMMIT
I, [2017-09-07 23:16:57 #7237] [task:46]  INFO -- DirectorJobRunner: Task 46 cancelled
D, [2017-09-07 23:16:57 #7237] [task:46] DEBUG -- DirectorJobRunner: (0.000092s) SELECT NULL
D, [2017-09-07 23:16:57 #7237] [task:46] DEBUG -- DirectorJobRunner: (0.000208s) SELECT * FROM "tasks" WHERE ("id" = 46) LIMIT 1
D, [2017-09-07 23:16:57 #7237] [task:46] DEBUG -- DirectorJobRunner: (0.000093s) SELECT NULL
D, [2017-09-07 23:16:57 #7237] [task:46] DEBUG -- DirectorJobRunner: (0.000049s) BEGIN
D, [2017-09-07 23:16:57 #7237] [task:46] DEBUG -- DirectorJobRunner: (0.000282s) UPDATE "tasks" SET "state" = 'cancelled', "timestamp" = '2017-09-07 23:16:57.201156+0000', "description" = 'create deployment', "result" = 'task cancelled', "output" = '/var/vcap/store/director/tasks/46', "checkpoint_time" = '2017-09-07 23:13:46.846333+0000', "type" = 'update_deployment', "username" = 'admin', "deployment_name" = 'consul-scale-down-5-to-3', "started_at" = NULL, "event_output" = '{"time":1504826217,"error":{"code":10001,"message":"Task 46 cancelled"}}
', "result_output" = '', "context_id" = '' WHERE ("id" = 46)
D, [2017-09-07 23:16:57 #7237] [task:46] DEBUG -- DirectorJobRunner: (0.002927s) COMMIT
I, [2017-09-07 23:16:57 #7237] []  INFO -- DirectorJobRunner: Task took 0.01756326 seconds to process.

Task 46 cancelled

Capturing task '46' output:
  Expected task '46' to succeed but state is 'cancelled'

Exit code 1
```

In all cases we have examined the time between these two log lines has been at least 3 minutes:

```
I, [2017-09-07T23:13:46.853554 #2834] [0x2ad3a7ebd110]  INFO -- TaskHelper: Enqueuing task: 46
I, [2017-09-07 23:16:57 #7237] []  INFO -- DirectorJobRunner: Looking for task with task id 46
```

Some tasks that wait that long *do* execute fine.  In those cases, we see that the `checkpoint_time` is updated, for example this task worked (see the line that has `UPDATE "tasks"`:

```
nm+dm $ bosh task 442 --debug | head -n 50
I, [2017-09-08T18:22:44.654271 #2834] [0x2ad3a7ebd110]  INFO -- TaskHelper: Director Version: 262.3.0
I, [2017-09-08T18:22:44.654364 #2834] [0x2ad3a7ebd110]  INFO -- TaskHelper: Enqueuing task: 442
I, [2017-09-08 18:25:40 #14416] []  INFO -- DirectorJobRunner: Looking for task with task id 442
D, [2017-09-08 18:25:40 #14416] [] DEBUG -- DirectorJobRunner: (0.000129s) SELECT NULL
D, [2017-09-08 18:25:40 #14416] [] DEBUG -- DirectorJobRunner: (0.000221s) SELECT * FROM "tasks" WHERE "id" = 442
I, [2017-09-08 18:25:40 #14416] []  INFO -- DirectorJobRunner: Found task #<Bosh::Director::Models::Task @values={:id=>442, :state=>"processing", :timestamp=>2017-09-08 18:22:44 UTC, :description=>"create deployment", :result=>nil, :output=>"/var/vcap/store/director/tasks/442", :checkpoint_time=>2017-09-08 18:22:44 UTC, :type=>"update_deployment", :username=>"admin", :deployment_name=>"consul-tls-key-rotation", :started_at=>nil, :event_output=>"", :result_output=>"", :context_id=>""}>
I, [2017-09-08 18:25:40 #14416] []  INFO -- DirectorJobRunner: Starting task: 442
I, [2017-09-08 18:25:40 #14416] [task:442]  INFO -- DirectorJobRunner: Creating job
D, [2017-09-08 18:25:40 #14416] [task:442] DEBUG -- DirectorJobRunner: (0.000094s) SELECT NULL
D, [2017-09-08 18:25:40 #14416] [task:442] DEBUG -- DirectorJobRunner: (0.000160s) SELECT * FROM "tasks" WHERE "id" = 442
D, [2017-09-08 18:25:40 #14416] [task:442] DEBUG -- DirectorJobRunner: (0.000078s) SELECT NULL
D, [2017-09-08 18:25:40 #14416] [task:442] DEBUG -- DirectorJobRunner: (0.000182s) SELECT * FROM "tasks" WHERE "id" = 442
I, [2017-09-08 18:25:40 #14416] [task:442]  INFO -- DirectorJobRunner: Performing task: #<Bosh::Director::Models::Task @values={:id=>442, :state=>"processing", :timestamp=>2017-09-08 18:22:44 UTC, :description=>"create deployment", :result=>nil, :output=>"/var/vcap/store/director/tasks/442", :checkpoint_time=>2017-09-08 18:22:44 UTC, :type=>"update_deployment", :username=>"admin", :deployment_name=>"consul-tls-key-rotation", :started_at=>nil, :event_output=>"", :result_output=>"", :context_id=>""}>
D, [2017-09-08 18:25:40 #14416] [task:442] DEBUG -- DirectorJobRunner: (0.000087s) SELECT NULL
D, [2017-09-08 18:25:40 #14416] [task:442] DEBUG -- DirectorJobRunner: (0.000049s) BEGIN
D, [2017-09-08 18:25:40 #14416] [task:442] DEBUG -- DirectorJobRunner: (0.000313s) UPDATE "tasks" SET "state" = 'processing', "timestamp" = '2017-09-08 18:25:40.025268+0000', "description" = 'create deployment', "result" = NULL, "output" = '/var/vcap/store/director/tasks/442', "checkpoint_time" = '2017-09-08 18:25:40.025439+0000', "type" = 'update_deployment', "username" = 'admin', "deployment_name" = 'consul-tls-key-rotation', "started_at" = '2017-09-08 18:25:40.025383+0000', "event_output" = '', "result_output" = '', "context_id" = '' WHERE ("id" = 442)
D, [2017-09-08 18:25:40 #14416] [task:442] DEBUG -- DirectorJobRunner: (0.001757s) COMMIT
I, [2017-09-08 18:25:40 #14416] [task:442]  INFO -- DirectorJobRunner: Reading deployment manifest
```

In the cases where that task is cancelled, the `checkpoint_time` is never updated and remains the timestamp of when the task was queued.

**Comment by dpb587-pivotal** [comment]
Is your director under very high load when this happens, or the database locked for a period around this time?

**Comment by mcwumbly** [comment]
I don't think the director was under high load in terms of RAM or CPU usage from what we could tell, but this theory from @jraqula in the [slack thread](https://cloudfoundry.slack.com/archives/C02HPPYQ2/p1504897693000476?thread_ts=1504894095.000256&cid=C02HPPYQ2) seems plausible:

> how's this for a theory: listing/checking a task will cause the state to be set to timeout if it's outside the 90s

Our tests [poll the director](https://github.com/pivotal-cf-experimental/bosh-test/blob/0a86d74cf998d2b7f256e319eef4d211b3c02e8d/bosh/client.go#L102-L130) to check the task status on an interval (that is 5 seconds by default).

**Comment by dpb587-pivotal** [comment]
Story [#152339383](https://www.pivotaltracker.com/story/show/152339383)

**Comment by dpb587-pivotal** [comment]
Closing - this was due to a race condition in task logic and was fixed in [v264.1+](https://github.com/cloudfoundry/bosh/releases/tag/v264.1).
